### PR TITLE
Improvements to make it work on Heroku.

### PR DESCRIPTION
**Pull request by calavera**
- Go 1.7.
- Logrus.
- Non-root user.
- Global variables.
- Extra configuration fixes.

### CODE REVIEW

File: api/api.go [hunk 1]
```diff
@@ -73,11 +74,22 @@ func (a *API) ListenAndServe(hostAndPort string) error {
 	return http.ListenAndServe(hostAndPort, a.handler)
 }
 
+func timeRequest(ctx context.Context, w http.ResponseWriter, r *http.Request) context.Context {
+	return context.WithValue(ctx, "_netlify_commerce_timing", time.Now())
+}
+
+func logHandler(ctx context.Context, wp mutil.WriterProxy, req *http.Request) {
+	start := ctx.Value("_netlify_commerce_timing").(time.Time)
+	logrus.Infof("method=%s path=%s status=%v duration=%v", req.Method, req.URL.Path, wp.Status(), time.Since(start))
```
Comment (rybit) [inline api/api.go]: It is fine to do this, but I would rather see `logrus.WithFields(xxxx).Info`. It will make for better JSON output

File: main.go [hunk 2]
```diff
@@ -8,7 +8,7 @@ import (
 )
 
 func main() {
-	cmd.InitCommandFlags()
+	cmd.InitCommands()
```
Comment (rybit) [inline main.go]: Instead of this I would like a `cmd.RootCmd()` that does the init. That way we don't forget to call it :)

File: conf/configuration.go [hunk 3]
```diff
@@ -47,6 +51,8 @@ type Configuration struct {
 
 // Load will construct the config from the file `config.json`
 func Load(configFile string) (*Configuration, error) {
+	viper.SetConfigType("json")
```
Comment (rybit) [inline conf/configuration.go]: why is this needed? it figures it out from the name afaik

File: conf/configuration.go [hunk 4]
```diff
@@ -60,30 +66,48 @@ func Load(configFile string) (*Configuration, error) {
...
 // This is supppper sad. It is b/c the marshal function doesn't work on nested
 // values. The overrides work, but the marshal won't discover them.
 // see: https://github.com/spf13/viper/issues/190
-func handleNested(config *Configuration) *Configuration {
+func handleNested(config *Configuration) (*Configuration, error) {
 	config.JWT.Secret = viper.GetString("jwt.secret")
 
 	config.DB.Driver = viper.GetString("db.driver")
 	config.DB.ConnURL = viper.GetString("db.url")
+	if config.DB.ConnURL == "" && os.Getenv("DATABASE_URL") != "" {
```
Comment (rybit) [inline conf/configuration.go]: this should be set-able by setting `NETLIFY_COMMERCE_DATABASE_URL` same as the `PORT`

File: api/api.go [hunk 5]
```diff
@@ -73,11 +74,22 @@ func (a *API) ListenAndServe(hostAndPort string) error {
...
+}
+
 // NewAPI instantiates a new REST API
 func NewAPI(config *conf.Configuration, db *gorm.DB, mailer *mailer.Mailer) *API {
 	api := &API{config: config, db: db, mailer: mailer, httpClient: &http.Client{}}
 	mux := kami.New()
+	mux.LogHandler = logHandler
```
Comment (rybit) [inline api/api.go]: I was looking at this, but in digging through the code I couldn't clearly see it called at the end. I might be wrong though.

Comment (calavera) [inline api/api.go]: it's called at the end

File: api/api.go (hunk 1 above)
Comment (calavera) [inline api/api.go]: agreed

File: main.go (hunk 2 above)
Comment (calavera) [inline main.go]: 👍

File: conf/configuration.go (hunk 4 above)
Comment (calavera) [inline conf/configuration.go]: it is, but heroku uses their own custom variables, that cannot be changed. Both are supported.

Comment (rybit) [inline conf/configuration.go]: le sad ~ as this is in a docker container couldn't we just map them over? Or am I misunderstanding the deployment process?

File: api/api.go (hunk 5 above)
Comment (rybit) [inline api/api.go]: In my PR I add a wrapper method that will do a lot of this b/c this doesn't call at the start or have a unique key. I prefer this to having a bunch of middleware's and storing things in the context. https://github.com/netlify/netlify-commerce/pull/6/files#diff-604f4192deea5adf9c159e5f88f64630R109

File: conf/configuration.go (hunk 3 above)
Comment (calavera) [inline conf/configuration.go]: This is what happens without this code when there are no configuration files, which is perfectly valid if you only use environment variables:

```

Comment (rybit) [inline conf/configuration.go]: good catch!

File: api/api.go (hunk 1 above)
Comment (calavera) [inline api/api.go]: On second thought, I like better the current format, I want that specific order, and logrus sorts alphabetically using Fields. Comparison:

Logrus:

```
INFO[0000] Netlify Commerce API started on: :8080       
INFO[0004]                                               duration=6.985µs method=GET path=/ status=200
INFO[0013]                                               duration=25.74µs method=GET path=/ status=200
```

Just Text:

```
INFO[0000] Netlify Commerce API started on: :8080       
INFO[0002] method=GET path=/ status=200 duration=10.025µs 
INFO[0005] method=GET path=/ status=200 duration=28.816µs 
```

Comment (rybit) [inline api/api.go]: could we do both? 

Also I think you can disable ordering on the logger ~ but it is a global thing

File: conf/configuration.go (hunk 4 above)
Comment (calavera) [inline conf/configuration.go]: everything is set via environment variables. There is no intermediate step to "map them over", and that's actually what this is doing, if you use `NETLIFY_COMMERCE_DATABASE_URL` great, if not, check those global variables.

File: api/api.go (hunk 1 above)
Comment (calavera) [inline api/api.go]: what's the benefit of using both? it's just text, I rather not have a huge white space in the middle of the log only for the shake of using a map to set values.

File: api/api.go (hunk 5 above)
Comment (calavera) [inline api/api.go]: To be honest, I really hope you don't plan on submitting that huge change as 1 single PR.

Comment (rybit) [inline api/api.go]: ..............I was. you're right ~ I'll break that up

File: api/api.go (hunk 1 above)
Comment (rybit) [inline api/api.go]: of using both is that you can then access the fields without parsing the text line. Otherwise it isn't very machine friendly.

File: conf/configuration.go (hunk 4 above)
Comment (rybit) [inline conf/configuration.go]: I meant that this is launched in a docker container right? So you could add something like:

```
ENV NETLIFY_COMMERCE_DB_URL=$DATABASE_URL
``` 

in the container.

**Comment by rybit** [comment]
:shipit: 

~ yay for rebase helllllll :)
